File: red/src/workflow.ts
import * as storage from './storage.ts';
import {finalize_backend} from 'colors-compute-red';
// Lifecycle graph and backend advice, the port of
// io.github.getcolors.clickhouse.workflow.

import { readPars, parName } from "red/cli";
import * as dryRun from "red/dry-run";
import { preflight } from "red/lifecycle";
import * as progress from "red/progress";
import * as tofu from "red/tofu";
import { adviceAdd, workflow, type Opts, type WireDecl, failed } from "red/workflow";
import * as tools from "./tools.ts";
import * as validate from "./validate.ts";

export const defaults: Opts = {
  "compute-prevent-destroy": true, "provider-compute": "hcloud",
  "provider-dns": "cloudflare", "provider-backend": "r2",
  workdir: ".colors",
};

export const lifecycleEvents = ["create", "delete"];

export async function startStep(
  opts: Opts,
  env: Record<string, string | undefined> = process.env,
): Promise<Opts> {
  return preflight(opts, {
    defaults,
    overlay: readPars,
    validators: [
      (_opts, environment) => validate.envErrors(environment),
      (current) => validate.stateErrors(current),
      (current, _environment, { event, real }) =>
        real && lifecycleEvents.includes(String(event))
          ? validate.secretErrors(current)
          : [],
      (current, _environment, { event, real }) =>
        real && event === "delete" && current["compute-prevent-destroy"]
          ? [`compute destruction is protected; set ${parName("compute-prevent-destroy")}=false to delete`]
          : [],
    ],
  }, env);
}

export function passStep(opts: Opts): Opts {
  return { ...opts, "red/exit": 0 };
}

export async function backendFinalizeStep(opts: Opts): Promise<Opts> {
  try {
    const result = await finalize_backend(opts,{...process.env,...storage.awsEnv(opts)});
    if (!['destroyed','absent','skipped'].includes(result.status)) throw Error('finalization refused');
    return {...opts,'red/exit':0};
  } catch { return {...opts,'red/exit':1,'red/err':'managed backend finalization refused; live or unowned state remains'}; }
}

export function wireFn(step: string, runOpts: Opts): WireDecl | undefined {
  if (runOpts["red/event"] === "delete") {
    const graph: Record<string, WireDecl> = {
      "clickhouse/start": [startStep, "clickhouse/load-infrastructure"],
      "clickhouse/load-infrastructure": [tools.loadInfrastructureStep, "clickhouse/dbt"],
      "clickhouse/dbt": [tools.dbtStep, "clickhouse/acceptance"],
      "clickhouse/acceptance": [tools.acceptanceStep, "clickhouse/ansible-cleanup"],
      "clickhouse/ansible-cleanup": [tools.ansibleCleanupStep, "clickhouse/ansible-local"],
      "clickhouse/ansible-local": [tools.ansibleLocalStep, "clickhouse/dns"],
      "clickhouse/dns": [tools.dnsStep, storage.managed(runOpts)?"clickhouse/storage":"clickhouse/infrastructure"],
      "clickhouse/storage": [storage.storageStep, "clickhouse/infrastructure"],
      "clickhouse/infrastructure": runOpts["s3-bucket-mode"]==="managed"?[tools.infrastructureStep,"clickhouse/backend-finalize"]:[tools.infrastructureStep],
      "clickhouse/backend-finalize": [backendFinalizeStep],
    };
    return graph[step];
  }
  const graph: Record<string, WireDecl> = {
    "clickhouse/start": [startStep, "clickhouse/infrastructure"],
    "clickhouse/infrastructure": [tools.infrastructureStep, storage.managed(runOpts)?"clickhouse/storage":"clickhouse/dns"],
    "clickhouse/storage": [storage.storageStep,"clickhouse/dns"],
    "clickhouse/dns": [tools.dnsStep, "clickhouse/ansible-local"],
    "clickhouse/ansible-local": [tools.ansibleLocalStep, "clickhouse/ansible-render"],
    "clickhouse/ansible-render": [tools.ansibleRenderStep, "clickhouse/wireguard"],
    "clickhouse/wireguard": [tools.wireguardStep,
                             "clickhouse/clickhouse-config", "clickhouse/metabase-config"],
    "clickhouse/clickhouse-config": [tools.clickhouseConfigStep, "clickhouse/dbt"],
    "clickhouse/metabase-config": [tools.metabaseConfigStep, "clickhouse/dbt"],
    "clickhouse/dbt": [tools.dbtStep, "clickhouse/acceptance"],
    "clickhouse/acceptance": [tools.acceptanceStep, runOpts["clickhouse-backup-bucket"]?"clickhouse/rehearsal":"clickhouse/drift"],
    "clickhouse/rehearsal": [tools.rehearsalStep, "clickhouse/drift"],
    "clickhouse/drift": [tools.driftStep],
  };
  return graph[step];
}

export function backendAdvice(tool: string) {
  return tofu.conventionalBackendAdvice({
    dir: (opts) => tools.toolDir(opts, tool),
    key: (opts) => `${opts.profile}/${tool}.tfstate`,
  });
}

export const sideEffecting = ['clickhouse/rehearsal','clickhouse/storage','clickhouse/backend-finalize','clickhouse/ansible-local',
  "clickhouse/infrastructure", "clickhouse/load-infrastructure",
  "clickhouse/dns", "clickhouse/wireguard", "clickhouse/clickhouse-config",
  "clickhouse/metabase-config", "clickhouse/ansible-cleanup", "clickhouse/dbt",
  "clickhouse/acceptance", "clickhouse/drift",
];

function create() {
  let wf = workflow({ start: "clickhouse/start", wireFn, nextFn: (_step, declared, opts) => failed(opts)||opts["clickhouse/already-destroyed"]?[]:opts["clickhouse/finalize-only"]?[["clickhouse/backend-finalize",{...opts,"clickhouse/finalize-only":false}]]:(declared??[]).map(step=>[step,opts] as [string,Opts]) });
  wf = progress.advise(wf);
  wf = dryRun.advise(wf, sideEffecting);
  for (const tool of tools.tofuTools) {
    wf = adviceAdd(wf, `clickhouse/${tool.slice("clickhouse-".length)}`, "before",
      `io.github.getcolors.clickhouse.workflow/backend-${tool}`, backendAdvice(tool));
  }
  return wf;
}

export const clickhouseWorkflow = create();

File: blue/src/package_clickhouse_blue/workflow.py
"""Lifecycle graph and backend advice, the port of
io.github.getcolors.clickhouse.workflow."""

from __future__ import annotations

from blue import dry_run, progress, tofu
from blue.cli import par_name, read_pars
from blue.lifecycle import preflight
from blue.workflow import advice_add, workflow, failed

from . import tools, validate

DEFAULTS = {"compute-prevent-destroy": True, "provider-compute": "hcloud",
            "provider-dns": "cloudflare", "provider-backend": "r2",
            "workdir": ".colors"}

LIFECYCLE_EVENTS = ("create", "delete")


async def start_step(opts: dict, env: dict | None = None) -> dict:
    return await preflight(
        opts, defaults=DEFAULTS, overlay=read_pars, env=env,
        validators=[
            lambda _o, e, _c: validate.env_errors(e),
            lambda o, _e, _c: validate.state_errors(o),
            lambda o, _e, c: (validate.secret_errors(o)
                              if c["real"] and c["event"] in LIFECYCLE_EVENTS else []),
            lambda o, _e, c: ([f"compute destruction is protected; set "
                               f"{par_name('compute-prevent-destroy')}=false to delete"]
                              if c["real"] and c["event"] == "delete"
                              and o.get("compute-prevent-destroy") else []),
        ])


def pass_step(opts: dict) -> dict:
    return {**opts, "blue/exit": 0}


def wire_fn(step, run_opts):
    if run_opts.get('blue/event') == 'delete':
        return {
            'clickhouse/start': (start_step, 'clickhouse/load-infrastructure'),
            'clickhouse/load-infrastructure': (tools.load_infrastructure_step, 'clickhouse/dbt'),
            'clickhouse/dbt': (tools.dbt_step, 'clickhouse/acceptance'),
            'clickhouse/acceptance': (tools.acceptance_step, 'clickhouse/ansible-cleanup'),
            'clickhouse/ansible-cleanup': (tools.ansible_cleanup_step, 'clickhouse/ansible-local'),
            'clickhouse/ansible-local': (tools.ansible_local_step, 'clickhouse/dns'),
            'clickhouse/dns': (tools.dns_step, 'clickhouse/infrastructure'),
            'clickhouse/infrastructure': (tools.infrastructure_step,),
        }.get(step)
    return {
        'clickhouse/start': (start_step, 'clickhouse/infrastructure'),
        'clickhouse/infrastructure': (tools.infrastructure_step, 'clickhouse/dns'),
        'clickhouse/dns': (tools.dns_step, 'clickhouse/ansible-local'),
        'clickhouse/ansible-local': (tools.ansible_local_step, 'clickhouse/ansible-render'),
        'clickhouse/ansible-render': (tools.ansible_render_step, 'clickhouse/wireguard'),
        'clickhouse/wireguard': (tools.wireguard_step, 'clickhouse/clickhouse-config', 'clickhouse/metabase-config'),
        'clickhouse/clickhouse-config': (tools.clickhouse_config_step, 'clickhouse/dbt'),
        'clickhouse/metabase-config': (tools.metabase_config_step, 'clickhouse/dbt'),
        'clickhouse/dbt': (tools.dbt_step, 'clickhouse/acceptance'),
        'clickhouse/acceptance': (tools.acceptance_step, 'clickhouse/drift'),
        'clickhouse/drift': (tools.drift_step,),
    }.get(step)


def backend_advice(tool: str):
    return tofu.conventional_backend_advice(
        dir=lambda o, tool=tool: tools.tool_dir(o, tool),
        key=lambda o, tool=tool: f"{o.get('profile')}/{tool}.tfstate")


side_effecting = ['clickhouse/ansible-local','clickhouse/infrastructure', 'clickhouse/load-infrastructure', 'clickhouse/dns',
    'clickhouse/wireguard', 'clickhouse/clickhouse-config', 'clickhouse/metabase-config',
    'clickhouse/ansible-cleanup', 'clickhouse/dbt', 'clickhouse/acceptance', 'clickhouse/drift']


def create_workflow():
    wf = workflow(start="clickhouse/start", wire_fn=wire_fn, next_fn=lambda step, successors, opts: [] if opts.get("clickhouse/already-destroyed") or failed(opts) else [(successor, opts) for successor in successors or []])
    wf = progress.advise(wf)
    wf = dry_run.advise(wf, side_effecting)
    for tool in tools.tofu_tools:
        wf = advice_add(wf, f"clickhouse/{tool[len('clickhouse-'):]}", "before",
                        f"io.github.getcolors.clickhouse.workflow/backend-{tool}",
                        backend_advice(tool))
    return wf


clickhouse_workflow = create_workflow()
